src/v1/sources/shopify/identityResolutionLayer.js

```diff
@@ -0,0 +1,87 @@
+const { v5 } = require('uuid');
+const { SHOPIFY_ADMIN_ONLY_EVENTS } = require('./config');
+const { getCartToken } = require('./commonUtils');
+const { generateUUID, isDefinedAndNotNull } = require('../../../v0/util');
+
+const idResolutionLayer = {
+  /**
+   * This function checks and returns rudderId from message if present
+   * returns null if not present or found
+   * @param {*} message
+   */
+  getRudderIdFromNoteAtrributes(noteAttributes, field) {
+    const rudderIdObj = noteAttributes.find((obj) => obj.name === field);
+    if (isDefinedAndNotNull(rudderIdObj)) {
+      return rudderIdObj.value;
+    }
+    return null;
+  },
+
+  /**
+   * This function retrieves anonymousId and sessionId in folowing steps:
+   * 1. Checks for `rudderAnonymousId`and `rudderSessionId in `note_atrributes`
+   * 2. Checks in dbData
+   * 3. This means we don't have `anonymousId` and hence events CAN NOT be stitched and we check for cartToken
+   *    a. if cartToken is available we return its hash value
+   *    b. else we check if the event is an SHOPIFY_ADMIN_ONLY_EVENT
+   *       -> if true we return `null`;
+   *       -> else we don't have any identifer (very edge case) we return `random anonymousId`
+   *    No Random SessionId is generated as its not a required field
+   * @param {*} message
+   * @param {*} metricMetadata
+   * @returns
+   */
+  getAnonymousIdAndSessionId(message, shopifyTopic, dbData = null) {
+    let anonymousId;
+    let sessionId;
+    const noteAttributes = message.properties?.note_attributes;
+    // Giving Priority to note_attributes to fetch rudderAnonymousId over Redis due to better efficiency
+    if (isDefinedAndNotNull(noteAttributes)) {
+      anonymousId = this.getRudderIdFromNoteAtrributes(noteAttributes, 'rudderAnonymousId');
+      sessionId = this.getRudderIdFromNoteAtrributes(noteAttributes, 'rudderSessionId');
+      if (anonymousId && sessionId) {
+        return { anonymousId, sessionId };
+      }
+    }
+
+    // falling back to cartToken mapping or its hash in case no rudderAnonymousId or rudderSessionId is found
+    const cartToken = getCartToken(message, shopifyTopic);
+    if (!isDefinedAndNotNull(cartToken)) {
+      if (SHOPIFY_ADMIN_ONLY_EVENTS.includes(message.event)) {
+        return { anonymousId, sessionId };
+      }
+      return {
+        anonymousId: isDefinedAndNotNull(anonymousId) ? anonymousId : generateUUID(),
+        sessionId,
+      };
+    }
+    anonymousId = anonymousId || dbData?.anonymousId;
+    sessionId = sessionId || dbData?.sessionId;
+    if (!isDefinedAndNotNull(anonymousId)) {
+      /* anonymousId or sessionId not found from db as well
+            Hash the id and use it as anonymousId (limiting 256 -> 36 chars) and sessionId is not sent as its not required field
+            */
+      anonymousId = v5(cartToken, v5.URL);
```

Comment:
Can we add a stat here? It will help us find how many events the redisDB search has failed.

Reply:
we already have that metric mentioned elsewhere